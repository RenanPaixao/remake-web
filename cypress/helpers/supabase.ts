import { createClient } from '@supabase/supabase-js'

const e2eSupabase = createClient(import.meta.env.VITE_SUPABASE_URL, import.meta.env.VITE_SUPABASE_SERVICE_ROLE_KEY, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  }
})

export { e2eSupabase }
