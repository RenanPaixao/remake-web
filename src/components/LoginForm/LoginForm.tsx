import { Field, FieldProps, Form, Formik, FormikErrors, FormikTouched } from 'formik'
import { Button, FormControl, FormErrorMessage, FormLabel, Input } from '@chakra-ui/react'
import { useNavigate } from 'react-router-dom'
import { supabase } from '../../utils/supabase.ts'
import * as Yup from 'yup'
import { useErrorToast } from '../../hooks/toast/useErrorToast.tsx'
import { useSuccessToast } from '../../hooks/toast/useSuccessToast.tsx'

const validationSchema = Yup.object({
  email: Yup.string().email('Invalid email address').required('Required'),
  password: Yup.string().required('Required').min(8, 'Password must be at least 8 characters')
})

type FormValues = Yup.InferType<typeof validationSchema>

export const LoginForm = () => {
  const errorToast = useErrorToast()
  const successToast = useSuccessToast()
  const navigate = useNavigate()

  /**
   * Submit the login form.
   * @param values
   */
  async function submitForm(values: FormValues): Promise<void> {
    const { error } = await supabase.auth.signInWithPassword({
      email: values.email,
      password: values.password
    })

    if (error) {
      errorToast({ description: error.message })
      console.error(error)
      return
    }

    successToast({ description: 'You have successfully logged in.' })

    navigate('/')
  }

  /**
   * Check if a field has an error.
   *
   * @param name - The name of the field.
   * @param touched - The object with all touched fields.
   * @param errors - The object with all errors.
   */
  function hasError(name: keyof FormValues, touched: FormikTouched<FormValues>, errors: FormikErrors<FormValues>): boolean {
    return !!touched[name] && !!errors[name]
  }

  return <Formik initialValues={{
    email: '',
    password: ''
  }}
  onSubmit={submitForm}
  validationSchema={validationSchema}
  >
    {({ isSubmitting, errors, touched }) => (
      <Form>
        <Field name="email">
          {({ field }: FieldProps) => (
            <FormControl isInvalid={hasError('email', touched, errors)} isRequired>
              <FormLabel htmlFor="email">Email</FormLabel>
              <Input {...field} id="email" placeholder="Email" />
              <FormErrorMessage>{errors.email}</FormErrorMessage>
            </FormControl>
          )}
        </Field>

        <Field name="password">
          {({ field }: FieldProps) => (
            <FormControl mt={4} isInvalid={hasError('password', touched, errors)} isRequired>
              <FormLabel htmlFor="password">Password</FormLabel>
              <Input type='password' {...field} id="password" placeholder="Password"/>
              <FormErrorMessage>{errors.password}</FormErrorMessage>
            </FormControl>
          )}
        </Field >

        <Button
          colorScheme='blue'
          color="white"
          role='button'
          mt={4}
          isLoading={isSubmitting}
          isDisabled={isSubmitting || Object.keys(errors).length > 0 }
          type="submit"
        >
          Login
        </Button>
      </Form>
    )}
  </Formik>
}
