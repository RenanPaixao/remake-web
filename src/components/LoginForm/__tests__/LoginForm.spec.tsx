import { LoginForm } from '../LoginForm.tsx'
import { customRender } from '../../../../tests/test-utils.tsx'
import { fireEvent, screen, waitFor } from '@testing-library/dom'
import { act } from '@testing-library/react'
import { expect } from 'vitest'
import { supabase } from '../../../utils/supabase.ts'

const navigateMock = vi.fn()
const supabaseLoginMock = vi.fn()
const toastMock = vi.fn()

supabase.auth.signInWithPassword = supabaseLoginMock
describe('LoginForm', () => {
  beforeEach(() => {
    vi.mock('react-router-dom', async importOriginal => {
      const actual: object = await importOriginal()
      return {
        ...actual,
        useNavigate: () => navigateMock
      }
    })

    vi.mock('@chakra-ui/react', async importOriginal => {
      const actual: object = await importOriginal()
      return {
        ...actual,
        useToast: () => toastMock
      }
    })
  })

  it('should render', async () => {
    const { container } = customRender(<LoginForm/>)

    expect(container).toMatchSnapshot()
  })

  it('should show validation errors', async () => {
    const { getByPlaceholderText, container } = customRender(<LoginForm/>)

    await fillForm('email', 'password')

    expect(container).toMatchSnapshot()

    const loginButton = screen.getByText('Login')

    expect(loginButton).toBeDisabled()

    await act(async () => {
      fireEvent.change(getByPlaceholderText('Email'), {
        target: {
          value: 'email@email.com'
        }
      })

      fireEvent.change(getByPlaceholderText('Password'), {
        target: {
          value: '12345678'
        }
      })
    })

    await waitFor(() => {
      expect(loginButton).toBeEnabled()
    }, {
      timeout: 2000
    })

    expect(container).toMatchSnapshot()
  })

  it('should not login successful', async () => {
    const error = new Error('auth error')
    supabaseLoginMock.mockReturnValue(Promise.resolve({
      error
    }))

    const { getByRole } = customRender(<LoginForm/>)

    await fillForm('wrong@email.com', '12345678')
    const loginButton = getByRole('button', { name: 'Login' })

    await waitFor(() => {
      expect(loginButton).toBeEnabled()
    }, {
      timeout: 2000
    })

    await act(async() => {
      fireEvent.click(loginButton)
    })

    await waitFor(() => {
      expect(toastMock).toHaveBeenCalledWith({
        title: 'Error',
        description: error.message,
        status: 'error',
        isClosable: true
      })
    })
  })

  it('should login successful', async () => {
    supabaseLoginMock.mockReturnValue(Promise.resolve({
      error: null
    }))

    const { getByRole } = customRender(<LoginForm/>)

    await fillForm('renan.paixao@email.com', '12345678')
    const loginButton = getByRole('button', { name: 'Login' })

    await waitFor(() => {
      expect(loginButton).toBeEnabled()
    }, {
      timeout: 2000
    })

    await act(async () => {
      fireEvent.click(loginButton)
    })

    await waitFor(() => {
      expect(toastMock).toHaveBeenCalledWith({
        title: 'Success',
        description: 'You have successfully logged in.',
        status: 'success',
        isClosable: true
      })
    },{
      timeout: 2000
    })

    expect(navigateMock).toHaveBeenCalledWith('/')
  })
})

const fillForm = async (email: string, password: string) => {
  const { getByPlaceholderText } = screen
  await act(async () => {
    fireEvent.change(getByPlaceholderText('Email'), {
      target: {
        value: email
      }
    })

    fireEvent.change(getByPlaceholderText('Password'), {
      target: {
        value: password
      }
    })
  })
}
