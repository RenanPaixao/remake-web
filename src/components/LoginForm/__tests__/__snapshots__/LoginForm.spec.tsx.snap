// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm > should render 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="email"
        id="field-:r0:-label"
      >
        Email
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-1tfjd1n"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required="true"
        class="chakra-input css-1cjy4zv"
        id="email"
        name="email"
        placeholder="Email"
        required=""
        value=""
      />
    </div>
    <div
      class="chakra-form-control css-1mn9ptq"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="password"
        id="field-:r1:-label"
      >
        Password
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-1tfjd1n"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required="true"
        class="chakra-input css-1cjy4zv"
        id="password"
        name="password"
        placeholder="Password"
        required=""
        type="password"
        value=""
      />
    </div>
    <button
      class="chakra-button css-13nworv"
      role="button"
      type="submit"
    >
      Login
    </button>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`LoginForm > should show validation errors 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="email"
        id="field-:r2:-label"
      >
        Email
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-1tfjd1n"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required="true"
        class="chakra-input css-1cjy4zv"
        id="email"
        name="email"
        placeholder="Email"
        required=""
        value="email"
      />
    </div>
    <div
      class="chakra-form-control css-1mn9ptq"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="password"
        id="field-:r3:-label"
      >
        Password
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-1tfjd1n"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required="true"
        class="chakra-input css-1cjy4zv"
        id="password"
        name="password"
        placeholder="Password"
        required=""
        type="password"
        value="password"
      />
    </div>
    <button
      class="chakra-button css-13nworv"
      disabled=""
      role="button"
      type="submit"
    >
      Login
    </button>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`LoginForm > should show validation errors 2`] = `
<div>
  <form
    action="#"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="email"
        id="field-:r2:-label"
      >
        Email
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-1tfjd1n"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required="true"
        class="chakra-input css-1cjy4zv"
        id="email"
        name="email"
        placeholder="Email"
        required=""
        value="email@email.com"
      />
    </div>
    <div
      class="chakra-form-control css-1mn9ptq"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="password"
        id="field-:r3:-label"
      >
        Password
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-1tfjd1n"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required="true"
        class="chakra-input css-1cjy4zv"
        id="password"
        name="password"
        placeholder="Password"
        required=""
        type="password"
        value="12345678"
      />
    </div>
    <button
      class="chakra-button css-13nworv"
      role="button"
      type="submit"
    >
      Login
    </button>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
